[SwiftUI] (macOS) Add support for disabling rubber-banding if the content size is less than that of the scroll view size

rr-codes opened the pull request:
#### 8bc5e1f0082a8335d68cfb2dd0913255e7cab7cd
<pre>
[SwiftUI] (macOS) Add support for disabling rubber-banding if the content size is less than that of the scroll view size
<a href="https://bugs.webkit.org/show_bug.cgi?id=287763">https://bugs.webkit.org/show_bug.cgi?id=287763</a>
<a href="https://rdar.apple.com/144929647">rdar://144929647</a>

Reviewed by Simon Fraser.

On iOS, clients may use the `alwaysBounceVertical` and `alwaysBounceHorizontal` properties on `WKWebView.scrollView`
to customize if rubber-banding/bouncing should be enabled if the content size is less than the scroll view size. The
default value of these properties in WebKit is `true`.

This PR brings support for this functionality to macOS by adding likewise, internal methods to WKWebView on macOS.

To facilitate this, the concept of &quot;rubber-bandable edges&quot; is slightly changed; instead of a particular edge being
either rubber-bandable or not, an edge now has three possible states: never rubber-bandable, always rubber-bandable,
and rubber-bandable only if the content size is greater than that of the scroll view size.

* LayoutTests/resources/ui-helper.js:
(window.UIHelper.async setAlwaysBounceVertical):
(window.UIHelper.async setAlwaysBounceHorizontal):
* Tools/DumpRenderTree/mac/UIScriptControllerMac.mm:
(WTR::setScrollBounceBehavior):
* Tools/TestRunnerShared/UIScriptContext/Bindings/UIScriptController.idl:
* Tools/TestRunnerShared/UIScriptContext/UIScriptController.h:
(WTR::UIScriptController::setScrollBounceBehavior):
(WTR::UIScriptController::setAlwaysBounceVertical):
(WTR::UIScriptController::setAlwaysBounceHorizontal):
* Tools/WebKitTestRunner/mac/UIScriptControllerMac.h:
* Tools/WebKitTestRunner/mac/UIScriptControllerMac.mm:
(WTR::UIScriptControllerMac::setAlwaysBounceVertical):
(WTR::UIScriptControllerMac::setAlwaysBounceHorizontal):

Test infrastructure.

* LayoutTests/scrollingcoordinator/mac/rubberbanding-always-bounce-short-content-expected.txt: Added.
* LayoutTests/scrollingcoordinator/mac/rubberbanding-always-bounce-short-content.html: Added.
* LayoutTests/scrollingcoordinator/mac/rubberbanding-bounce-based-on-size-short-content-expected.txt: Added.
* LayoutTests/scrollingcoordinator/mac/rubberbanding-bounce-based-on-size-short-content.html: Added.
* LayoutTests/scrollingcoordinator/mac/rubberbanding-bounce-based-on-size-tall-content-expected.txt: Added.
* LayoutTests/scrollingcoordinator/mac/rubberbanding-bounce-based-on-size-tall-content.html: Added.

Several new layout tests are added.

* Source/WebCore/page/scrolling/ScrollingTree.cpp:
(WebCore::ScrollingTree::setClientAllowedMainFrameRubberBandableEdges):
(WebCore::ScrollingTree::clientAllowsMainFrameRubberBandingOnSide):
(WebCore::ScrollingTree::willWheelEventStartSwipeGesture):
* Source/WebCore/page/scrolling/ScrollingTree.h:

Change the relevant boolean values to the new multiple-state enum value.

* Source/WebCore/page/scrolling/ScrollingTreeScrollingNode.cpp:
(WebCore::ScrollingTreeScrollingNode::shouldRubberBandOnSide const):

This contains the actual logic that resolves the three possible states into a binary decision of if
an edge should rubber-band while consulting if the content is scrollable.

* Source/WebCore/platform/ScrollTypes.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/platform/ScrollTypes.h:
* Source/WebCore/platform/ScrollTypes.serialization.in:
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Add a new enum type to represent the new multiple states.

* Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:

Unified sources fix.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/API/mac/WKWebViewMac.mm:
(-[WKWebView _alwaysBounceVertical]):
(-[WKWebView _setAlwaysBounceVertical:]):
(-[WKWebView _alwaysBounceHorizontal]):
(-[WKWebView _setAlwaysBounceHorizontal:]):

Exposes the new functionality in the internal WKWebView header interface.

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.cpp:
(WebKit::RemoteScrollingCoordinatorProxy::handleWheelEvent):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeEventDispatcher.cpp:
(WebKit::RemoteLayerTreeEventDispatcher::handleWheelEvent):
(WebKit::RemoteLayerTreeEventDispatcher::scrollingThreadHandleWheelEvent):
(WebKit::RemoteLayerTreeEventDispatcher::determineWheelEventProcessing):
(WebKit::RemoteLayerTreeEventDispatcher::handleSyntheticWheelEvent):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeEventDispatcher.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.mm:
(WebKit::RemoteScrollingCoordinatorProxyMac::handleWheelEvent):

Change the relevant boolean values to the new multiple-state enum value.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::resolvedScrollBounceBehaviorEdges):
(WebKit::WebPageProxy::handleWheelEvent):
(WebKit::WebPageProxy::continueWheelEventHandling):
(WebKit::WebPageProxy::sendWheelEvent):
(WebKit::WebPageProxy::alwaysBounceVertical const):
(WebKit::WebPageProxy::setAlwaysBounceVertical):
(WebKit::WebPageProxy::alwaysBounceHorizontal const):
(WebKit::WebPageProxy::setAlwaysBounceHorizontal):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxyInternals.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::colorSpace):
(WebKit::WebViewImpl::alwaysBounceVertical):
(WebKit::WebViewImpl::setAlwaysBounceVertical):
(WebKit::WebViewImpl::alwaysBounceHorizontal):
(WebKit::WebViewImpl::setAlwaysBounceHorizontal):
* Source/WebKit/WebProcess/WebPage/EventDispatcher.cpp:
(WebKit::EventDispatcher::internalWheelEvent):
(WebKit::EventDispatcher::wheelEvent):
(WebKit::EventDispatcher::handleSyntheticWheelEvent):
* Source/WebKit/WebProcess/WebPage/EventDispatcher.h:
* Source/WebKit/WebProcess/WebPage/EventDispatcher.messages.in:
* Source/WebKit/WebProcess/WebPage/MomentumEventDispatcher.cpp:
(WebKit::MomentumEventDispatcher::handleWheelEvent):
* Source/WebKit/WebProcess/WebPage/MomentumEventDispatcher.h:

Change the relevant boolean values to the new multiple-state enum value, and propogate the values from the UI process to the web process.

Canonical link: <a href="https://commits.webkit.org/290494@main">https://commits.webkit.org/290494@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/97c787a41fb05fa069a5d406b74e8c26c93af996

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/90221 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/9750 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/45144 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/95222 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/40997 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/92273 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/10138 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/18063 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/69457 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27059 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/93222 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/7774 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/81850 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49817 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/89722 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/7497 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/36223 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/40128 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/77828 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/37278 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/97047 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/17409 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/12807 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/78459 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/17666 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/77674 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/77665 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19174 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22121 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/20726 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/10668 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/17419 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/22746 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/17160 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/20612 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/18944 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/WebProcess/WebPage/MomentumEventDispatcher.h:
```diff
@@ -34,6 +34,7 @@
 #include <WebCore/FloatSize.h>
 #include <WebCore/PageIdentifier.h>
 #include <WebCore/RectEdges.h>
+#include <WebCore/ScrollTypes.h>
```
Does `RectEdges<T>` need the type definition? Maybe you can get away with declaring `ScrollBounceBehavior` here?

smfr commented on Source/WebCore/platform/ScrollTypes.h:
```diff
@@ -151,6 +151,13 @@ enum class ScrollElasticity : uint8_t {
 };
 
+// Determines if rubber-banding should be enabled if the content size is less than the scroll view size.
+enum class ScrollBounceBehavior : uint8_t {
```
Let's call this `RubberBandingBehavior` so it matches all the variable names. It's OK to have a disconnect at the API boundary (better than the type not matching variable names everywhere).

smfr commented on Source/WebCore/page/scrolling/ScrollingTree.cpp:
```diff
@@ -953,13 +953,13 @@ bool ScrollingTree::willWheelEventStartSwipeGesture(const PlatformWheelEvent& wh
     Locker locker { m_swipeStateLock };
 
-    if (wheelEvent.deltaX() > 0 && m_swipeState.mainFramePinnedState.left() && !m_swipeState.clientAllowedRubberBandableEdges.left())
+    if (wheelEvent.deltaX() > 0 && m_swipeState.mainFramePinnedState.left() && m_swipeState.clientAllowedRubberBandableEdges.left() == ScrollBounceBehavior::Never)
```
if it's "based on size" and the size means it's not scrollable, should that allow swipe?

smfr commented on Source/WebCore/page/scrolling/ScrollingTreeScrollingNode.cpp:
```diff
@@ -142,7 +142,9 @@ bool ScrollingTreeScrollingNode::shouldRubberBandOnSide(BoxSide side, RectEdges<
         return false;
 
-    if (isRootNode() && !scrollingTree()->clientAllowsMainFrameRubberBandingOnSide(side))
+    auto allowsMainFrameRubberBanding = scrollingTree()->clientAllowsMainFrameRubberBandingOnSide(side);
```
The `auto` hides the fact that the value isn't a bool, making "allows" confusing. Call it `mainFrameRubberBandingBehavior`.

rr-codes committed:
[SwiftUI] (macOS) Add support for disabling rubber-banding if the content size is less than that of the scroll view size
https://bugs.webkit.org/show_bug.cgi?id=287763
rdar://144929647

Reviewed by Simon Fraser.

On iOS, clients may use the `alwaysBounceVertical` and `alwaysBounceHorizontal` properties on `WKWebView.scrollView`
to customize if rubber-banding/bouncing should be enabled if the content size is less than the scroll view size. The
default value of these properties in WebKit is `true`.

This PR brings support for this functionality to macOS by adding likewise, internal methods to WKWebView on macOS.

To facilitate this, the concept of "rubber-bandable edges" is slightly changed; instead of a particular edge being
either rubber-bandable or not, an edge now has three possible states: never rubber-bandable, always rubber-bandable,
and rubber-bandable only if the content size is greater than that of the scroll view size.

* LayoutTests/resources/ui-helper.js:
(window.UIHelper.async setAlwaysBounceVertical):
(window.UIHelper.async setAlwaysBounceHorizontal):
* Tools/DumpRenderTree/mac/UIScriptControllerMac.mm:
(WTR::setScrollBounceBehavior):
* Tools/TestRunnerShared/UIScriptContext/Bindings/UIScriptController.idl:
* Tools/TestRunnerShared/UIScriptContext/UIScriptController.h:
(WTR::UIScriptController::setScrollBounceBehavior):
(WTR::UIScriptController::setAlwaysBounceVertical):
(WTR::UIScriptController::setAlwaysBounceHorizontal):
* Tools/WebKitTestRunner/mac/UIScriptControllerMac.h:
* Tools/WebKitTestRunner/mac/UIScriptControllerMac.mm:
(WTR::UIScriptControllerMac::setAlwaysBounceVertical):
(WTR::UIScriptControllerMac::setAlwaysBounceHorizontal):

Test infrastructure.

* LayoutTests/scrollingcoordinator/mac/rubberbanding-always-bounce-short-content-expected.txt: Added.
* LayoutTests/scrollingcoordinator/mac/rubberbanding-always-bounce-short-content.html: Added.
* LayoutTests/scrollingcoordinator/mac/rubberbanding-bounce-based-on-size-short-content-expected.txt: Added.
* LayoutTests/scrollingcoordinator/mac/rubberbanding-bounce-based-on-size-short-content.html: Added.
* LayoutTests/scrollingcoordinator/mac/rubberbanding-bounce-based-on-size-tall-content-expected.txt: Added.
* LayoutTests/scrollingcoordinator/mac/rubberbanding-bounce-based-on-size-tall-content.html: Added.

Several new layout tests are added.

* Source/WebCore/page/scrolling/ScrollingTree.cpp:
(WebCore::ScrollingTree::setClientAllowedMainFrameRubberBandableEdges):
(WebCore::ScrollingTree::clientAllowsMainFrameRubberBandingOnSide):
(WebCore::ScrollingTree::willWheelEventStartSwipeGesture):
* Source/WebCore/page/scrolling/ScrollingTree.h:

Change the relevant boolean values to the new multiple-state enum value.

* Source/WebCore/page/scrolling/ScrollingTreeScrollingNode.cpp:
(WebCore::ScrollingTreeScrollingNode::shouldRubberBandOnSide const):

This contains the actual logic that resolves the three possible states into a binary decision of if
an edge should rubber-band while consulting if the content is scrollable.

* Source/WebCore/platform/ScrollTypes.cpp:
(WebCore::operator<<):
* Source/WebCore/platform/ScrollTypes.h:
* Source/WebCore/platform/ScrollTypes.serialization.in:
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Add a new enum type to represent the new multiple states.

* Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:

Unified sources fix.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/API/mac/WKWebViewMac.mm:
(-[WKWebView _alwaysBounceVertical]):
(-[WKWebView _setAlwaysBounceVertical:]):
(-[WKWebView _alwaysBounceHorizontal]):
(-[WKWebView _setAlwaysBounceHorizontal:]):

Exposes the new functionality in the internal WKWebView header interface.

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.cpp:
(WebKit::RemoteScrollingCoordinatorProxy::handleWheelEvent):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeEventDispatcher.cpp:
(WebKit::RemoteLayerTreeEventDispatcher::handleWheelEvent):
(WebKit::RemoteLayerTreeEventDispatcher::scrollingThreadHandleWheelEvent):
(WebKit::RemoteLayerTreeEventDispatcher::determineWheelEventProcessing):
(WebKit::RemoteLayerTreeEventDispatcher::handleSyntheticWheelEvent):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeEventDispatcher.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.mm:
(WebKit::RemoteScrollingCoordinatorProxyMac::handleWheelEvent):

Change the relevant boolean values to the new multiple-state enum value.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::resolvedScrollBounceBehaviorEdges):
(WebKit::WebPageProxy::handleWheelEvent):
(WebKit::WebPageProxy::continueWheelEventHandling):
(WebKit::WebPageProxy::sendWheelEvent):
(WebKit::WebPageProxy::alwaysBounceVertical const):
(WebKit::WebPageProxy::setAlwaysBounceVertical):
(WebKit::WebPageProxy::alwaysBounceHorizontal const):
(WebKit::WebPageProxy::setAlwaysBounceHorizontal):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxyInternals.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::colorSpace):
(WebKit::WebViewImpl::alwaysBounceVertical):
(WebKit::WebViewImpl::setAlwaysBounceVertical):
(WebKit::WebViewImpl::alwaysBounceHorizontal):
(WebKit::WebViewImpl::setAlwaysBounceHorizontal):
* Source/WebKit/WebProcess/WebPage/EventDispatcher.cpp:
(WebKit::EventDispatcher::internalWheelEvent):
(WebKit::EventDispatcher::wheelEvent):
(WebKit::EventDispatcher::handleSyntheticWheelEvent):
* Source/WebKit/WebProcess/WebPage/EventDispatcher.h:
* Source/WebKit/WebProcess/WebPage/EventDispatcher.messages.in:
* Source/WebKit/WebProcess/WebPage/MomentumEventDispatcher.cpp:
(WebKit::MomentumEventDispatcher::handleWheelEvent):
* Source/WebKit/WebProcess/WebPage/MomentumEventDispatcher.h:

Change the relevant boolean values to the new multiple-state enum value, and propogate the values from the UI process to the web process.

Canonical link: https://commits.webkit.org/290494@main
